File: twao/Player.java
package twao;

import java.util.List;
import java.util.LinkedList;


public class Player {
    private final String nickname;
    private int numberOfVillages = 0;
    private int numberOfNobles;
    private final List<VillageAssignment> offAssignments = new LinkedList<>();
    private final List<VillageAssignment> fakeAssignments = new LinkedList<>();
    private final List<VillageAssignment> nobleAssignments = new LinkedList<>();
    private final List<VillageAssignment> fakeNobleAssignments = new LinkedList<>();

    public Player(String nickname, int numberOfNobles) {
        this.nickname = nickname;
        this.numberOfNobles = numberOfNobles;
    }

    /**
     * -----------------------------------------------------------
     * getters and setters
     * -----------------------------------------------------------
     */

    public String getNickname() { return nickname; }

    public int getNumberOfVillages() { return numberOfVillages; }

    public void increaseNumberOfVillages() { numberOfVillages++; }

    public void decreaseNumberOfVillaes() { numberOfVillages--; }

    public boolean hasNoble() { return (numberOfNobles > 0) ? true : false; }

    public void delegateNoble() { numberOfNobles--; }

    public int getNumberOfNobles() { return numberOfNobles; }

    public synchronized List<VillageAssignment> getOffAssignments() { return offAssignments; }

    public synchronized List<VillageAssignment> getFakeAssignments() { return fakeAssignments; }

    public synchronized List<VillageAssignment> getNobleAssignments() { return nobleAssignments; }

    public synchronized List<VillageAssignment> getFakeNobleAssignments() { return fakeNobleAssignments; }
}

File: src/main/java/io/github/stiv3ns/twactionorganizer/twao/Player.kt
package io.github.stiv3ns.twactionorganizer.twao

import kotlinx.serialization.Serializable

@Serializable
data class Player(val nickname: String) {
    var numberOfNobles: Int = 0
        private set
    var numberOfVillages: Int = 0
        private set

    constructor(nickname: String, numberOfNobles: Int = 0) : this(nickname) {
        this.numberOfNobles = numberOfNobles
    }

    val offAssignmentsCopy: List<VillageAssignment>
        get() = offAssignments.toList()

    val fakeAssignmentsCopy: List<VillageAssignment>
        get() = fakeAssignments.toList()

    val nobleAssignmentsCopy: List<VillageAssignment>
        get() = nobleAssignments.toList()

    val fakeNobleAssignmentsCopy: List<VillageAssignment>
        get() = fakeNobleAssignments.toList()

    private val offAssignments       = mutableListOf<VillageAssignment>()
    private val fakeAssignments      = mutableListOf<VillageAssignment>()
    private val nobleAssignments     = mutableListOf<VillageAssignment>()
    private val fakeNobleAssignments = mutableListOf<VillageAssignment>()


    fun hasNoble(): Boolean = (numberOfNobles > 0)

    fun delegateNoble() {
        numberOfNobles--
    }


    fun registerVillage() {
        numberOfVillages++
    }

    fun unregisterVillage() {
        numberOfVillages--
    }


    fun putOffAssignment(assignment: VillageAssignment) {
        offAssignments.add(assignment)
    }

    @Synchronized fun putFakeAssignment(assignment: VillageAssignment) {
        fakeAssignments.add(assignment)
    }

    fun putNobleAssignment(assignment: VillageAssignment) {
        nobleAssignments.add(assignment)
    }

    fun putFakeNobleAssignment(assignment: VillageAssignment) {
        fakeNobleAssignments.add(assignment)
    }
}